src/Chainweb/Pact/PactService/ExecBlock.hs

```diff
@@ -376,48 +378,58 @@ applyPactCmds
     -> Vector ChainwebTransaction
     -> Miner
     -> ModuleCache
+    -> Maybe P.Gas
     -> PactServiceM cas (Vector (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value])))
-applyPactCmds isGenesis env cmds miner mc =
-    V.fromList . toList . sfst <$> V.foldM f (T2 mempty mc) cmds
+applyPactCmds isGenesis env cmds miner mc blockGas = do
+    evalStateT (V.mapM go cmds) (T2 mc blockGas)
   where
-    f  (T2 dl mcache) cmd = applyPactCmd isGenesis env cmd miner mcache dl
-
--- | Apply a single Pact command
-applyPactCmd
-    :: Bool
-    -> P.PactDbEnv p
-    -> ChainwebTransaction
-    -> Miner
-    -> ModuleCache
-    -> DList (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))
-    -> PactServiceM cas
-       (T2 (DList (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))) ModuleCache)
-applyPactCmd isGenesis dbEnv cmdIn miner mcache dl = do
-    logger <- view psLogger
-    gasModel <- view psGasModel
-    v <- view psVersion
-
-    handle onBuyGasFailure $ do
-      T2 result mcache' <- if isGenesis
-        then liftIO $! applyGenesisCmd logger dbEnv P.noSPVSupport (payloadObj <$> cmdIn)
-        else do
-          pd <- getTxContext (publicMetaOf $ payloadObj <$> cmdIn)
-          spv <- use psSpvSupport
-          liftIO $! applyCmd v logger dbEnv miner (gasModel pd) pd spv cmdIn mcache
-
-      when isGenesis $
-        updateInitCache mcache'
-
-      unless isGenesis $ debugResult "applyPactCmd" result
-
-      cp <- getCheckpointer
-      -- mark the tx as processed at the checkpointer.
-      liftIO $ _cpRegisterProcessedTx cp (P._cmdHash cmdIn)
-      pure $! T2 (DL.snoc dl (Right result)) mcache'
-
-  where
-    onBuyGasFailure (BuyGasFailure f) = pure $! T2 (DL.snoc dl (Left f)) mcache
-    onBuyGasFailure e = throwM e
+    go cmd = StateT $ \(T2 mcache blockGasRemaining) -> do
```

Comment:
Hmm ... this used to at least scope what was happening at the single pact command level well which made for more legible code. At a minimum rename `go` to `applyPactCmd` but consider retaining toplevel function. Especially as it makes this diff hard to read.